File: Data Structures & Algorithms/binary-tree-diameter/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    #longest = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        self.longest = 0
        
        self.maxHeight(root)

        return self.longest
        
    def maxHeight(self, root: Optional[TreeNode]) -> int:

        if not root:
            return 0
        
        leftH, rightH = self.maxHeight(root.left), self.maxHeight(root.right)

        self.longest = max(self.longest, leftH + rightH)

        return 1 + max(leftH, rightH)

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-6.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        return serialize(root).contains(serialize(subRoot));
    }

    public String serialize(TreeNode node) {
        if(node == null) {return "#";}
        return String.format(".{%d}.{%s}.{%s}", node.val, serialize(node.left), serialize(node.right));
    }
}
